feat: Update Logger#on_emit to create LogRecords

kaylareopelle opened the pull request:
This PR creates a `LogRecord` class that is read/writeable to start with, and becomes read-only when it transitions to `LogRecordData`. 

Additionally, it provides the SDK implementation of the Logs Bridge API `Logger#on_emit` method to emit `LogRecord`s to the processing pipeline.

Relates to the following specs:
* [Log Record Definition](https://opentelemetry.io/docs/specs/otel/logs/data-model/#log-and-event-record-definition)
* [Readable Log Record](https://opentelemetry.io/docs/specs/otel/logs/sdk/#readablelogrecord) 
* [Read Write Log Record](https://opentelemetry.io/docs/specs/otel/logs/sdk/#readwritelogrecord)
* [Logger Operations: Emit a log record](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord)

Resolves: https://github.com/open-telemetry/opentelemetry-ruby/issues/1484

kaylareopelle commented on logs_sdk/lib/opentelemetry/sdk/logs/logger.rb:
```diff
@@ -25,6 +27,56 @@ def initialize(name, version, logger_provider)
+        # @api public
+        def on_emit(timestamp: nil,
+                 observed_timestamp: nil,
+                 span_context: nil, # or should this just be context? like in the API?
```
Discussion here: https://github.com/open-telemetry/opentelemetry-ruby/pull/1611/files#r1513443549

kaylareopelle commented on logs_sdk/lib/opentelemetry/sdk/logs/logger.rb:
```diff
@@ -25,6 +27,56 @@ def initialize(name, version, logger_provider)
+                                     attributes: attributes,
+                                     logger: self)
+
+          logger_provider.instance_variable_get(:@log_record_processors).each do |processor|
+            processor.on_emit(log_record, span_context)
+          end
```
I think this might not be the best implementation approach. I'd like some feedback here.

This made more sense when we were exposing `log_record_processors` with an `attr_reader`. However, we decided not to do that because [it might not be safe for concurrent execution.](https://github.com/open-telemetry/opentelemetry-ruby/pull/1517/commits/de7f36ce8fa8eab285dea579afe69e2501bfb635#r1346607101)

The [Span gets its processors defined as an attribute during creation](https://github.com/open-telemetry/opentelemetry-ruby/blob/98c629edbc6445305bed65c4f4306431d683b474/sdk/lib/opentelemetry/sdk/trace/tracer_provider.rb#L145), but it's also created by the TracerProvider rather than the Tracer. The Tracer is equivalent to the Logger in this scenario.

Also unlike Span, there isn't a `finish` operation separate from `start`, so [`emit` both creates the log record and sends it to the processing pipeline.](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord)

kaylareopelle replied:
That's a great idea, @xuan-cao-swi. Incorporated in 1f9f9027219888422d638fc4c24b6be15f52dd78 (along with other context-related updates).

kaylareopelle commented on logs_api/lib/opentelemetry/logs/logger.rb:
```diff
@@ -12,36 +12,45 @@ class Logger
       #
       # @api public
-      def emit(
+      def on_emit(
```
When we renamed the method in https://github.com/open-telemetry/opentelemetry-ruby/pull/1517, I forgot to go back and change the references in the API.

robertlaurin commented on logs_sdk/lib/opentelemetry/sdk/logs/log_record.rb:
```diff
@@ -0,0 +1,99 @@
+        )
+          @timestamp = timestamp
+          @observed_timestamp = observed_timestamp || timestamp || Process.clock_gettime(Process::CLOCK_REALTIME, :nanosecond)
+          @span_context = span_context
```
Where is `span_context` coming from here?

kaylareopelle replied:
Fixed in: 647e0c930921fec0b14282ce4ed7ca920cc083f7

robertlaurin commented on logs_sdk/lib/opentelemetry/sdk/logs/log_record.rb:
```diff
@@ -0,0 +1,97 @@
+          @span_id = span_id
+          @trace_flags = trace_flags
+          @resource = logger&.resource
+          @instrumentation_scope = logger&.instrumentation_scope
```
Is it possible for a log record to get created from an api logger? Otherwise I would expect instrumentation scope to always be present (do we need the safe operator here?).

kaylareopelle replied:
@robertlaurin - From my reading, it seems like it's in-spec to allow an API Logger to emit a log record. 

Here's the part of the spec I'm working from: https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord 

What are your thoughts?

kaylareopelle replied:
This isn't really relevant anymore after the refactor in ab570713d783b1dac3edea013c2d6f929cc03549

robertlaurin commented on logs_sdk/lib/opentelemetry/sdk/logs/log_record.rb:
```diff
@@ -0,0 +1,97 @@
+          @trace_id = trace_id
+          @span_id = span_id
+          @trace_flags = trace_flags
+          @resource = logger&.resource
```
I know there's a method added to the sdk logger here to support this but it still reads like `logger.logger_provider.resource` to me. 

I wonder if there's any merit in following the same pattern we did with the tracer https://github.com/open-telemetry/opentelemetry-ruby/blob/main/sdk/lib/opentelemetry/sdk/trace/tracer.rb#L35

kaylareopelle replied:
That's a great point, following that approach cleans up the code nicely. Fixed in ab570713d783b1dac3edea013c2d6f929cc03549

xuan-cao-swi commented on logs_sdk/lib/opentelemetry/sdk/logs/log_record_data.rb:
```diff
@@ -0,0 +1,43 @@
+                                 :resource,              # optional OpenTelemetry::SDK::Resources::Resource
+                                 :instrumentation_scope, # OpenTelemetry::SDK::InstrumentationScope
+                                 :total_recorded_attributes) do # Integer
+        def unix_nano_timestamp
```
nit: is that possible to do `alias_method :unix_nano_observed_timestamp, :unix_nano_timestamp`?

kaylareopelle replied:
Good question! The current setup [follows the pattern for export helpers in the SpanData class](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/sdk/lib/opentelemetry/sdk/trace/span_data.rb#L34-L50).

They're not exactly aliasable because the `unix_nano_timestamp` method references the `timestamp` attribute on the LogRecordData Struct, whereas the other method uses the `observed_timestamp` attribute. I could refactor it to accept an argument, but then we'd be breaking from the pattern in SpanData. 

I'd be curious to hear what approach the maintainers prefer we take.

kaylareopelle replied:
Whoops, commented on this before I saw Francis' comment. These methods will be removed.

fbogsany commented on logs_sdk/lib/opentelemetry/sdk/logs/log_record_data.rb:
```diff
@@ -0,0 +1,43 @@
+  module SDK
+    module Logs
+      # LogRecordData is a Struct containing {LogRecord} data for export.
+      LogRecordData = Struct.new(:timestamp,             # optional Integer nanoseconds since Epoch
+                                 :observed_timestamp,    # Integer nanoseconds since Epoch
```
The `timestamp` and `observed_timestamp` are documented here to be "Integer nanoseconds since Epoch", so there should be no need for the `unix_nano_timestamp` and `unix_nano_observed_timestamp` methods below.

kaylareopelle replied:
Great point. Fixed in 0fb4606e20c8d71d4f2c47a71ac59898f4deb2e1

fbogsany commented on logs_sdk/lib/opentelemetry/sdk/logs/log_record.rb:
```diff
@@ -0,0 +1,97 @@
+
+        # Creates a new {LogRecord}.
+        #
+        # @param [optional Float, Time] timestamp Time when the event occurred.
+        # @param [optional Float, Time] observed_timestamp Time when the event
+        #   was observed by the collection system. If nil, will first attempt
+        #   to set to `timestamp`. If `timestamp` is nil, will set to
+        #   `Process.clock_gettime(Process::CLOCK_REALTIME, :nanosecond)`.
```
Why do these accept `Float`? They should only accept `Time`, like corresponding methods in the tracing API. This makes it easier to ensure the correctness of conversion to "Integer nanoseconds since Epoch" internally, and reduces confusion about the units.

kaylareopelle replied:
I got a little too overzealous when trying to adhere to the spec. The datatype in the Logs Data Model was listed as "Type: Timestamp, uint64 nanoseconds since Unix epoch." Since the timestamp could be efficiently accessed using `Process.clock_gettime(Process::CLOCK_REALTIME, :nanosecond)`, I decided to add it and make it the default.

Based on these questions, I decided to remove the `Float` option, go all in on the `Time` object and transform it into a nanosecond `Integer` in `LogRecord#to_log_record_data`. 

Fixes across two commits: 
* d3607592e8127ec0e6f0ca719a7d52649e7b5a5e 
* 0fb4606e20c8d71d4f2c47a71ac59898f4deb2e1

fbogsany commented on logs_api/lib/opentelemetry/logs/logger.rb:
```diff
@@ -12,36 +12,45 @@ class Logger
       #
-      # @param timestamp [optional Float, Time] Time in nanoseconds since Unix
+      # @param [optional Float, Time] timestamp Time in nanoseconds since Unix
       #   epoch when the event occurred measured by the origin clock, i.e. the
       #   time at the source.
-      # @param observed_timestamp [optional Float, Time] Time in nanoseconds
+      # @param [optional Float, Time] observed_timestamp Time in nanoseconds
       #   since Unix epoch when the event was observed by the collection system.
       #   Intended default: Process.clock_gettime(Process::CLOCK_REALTIME)
```
The units are unclear here. If we accept nanoseconds, I'd expect that to be Integer rather than Float, and if we accept `Time`, then the units are not relevant. IMO we should only accept `Time`, and we should convert that to "Integer nanoseconds since Epoch" internally.

Related: documenting the `Process.clock_gettime(Process::CLOCK_REALTIME)` as the default increases confusion about the units (and precision).

kaylareopelle replied:
Great points, I agree. Addressed here: https://github.com/open-telemetry/opentelemetry-ruby/pull/1611#discussion_r1678505898

kaylareopelle commented:
@robertlaurin, @fbogsany - Thanks for your feedback! I've been able to return to this project and addressed your comments. This PR is ready for another review.

mwear reviewed (approved):
LGTM. @fbogsany can you confirm that your changes have been addressed?

mwear commented:
I'll plan to merge this EOD Thursday if there aren't any objections.
